lib/openshift/service.rb

```diff
@@ -80,6 +80,14 @@ def ip(user: nil, cached: true, quiet: false)
       spec = raw_resource(user: user, cached: cached, quiet: quiet).dig('spec')
       return (spec.dig('portalIP') || spec.dig('clusterIP'))
     end
+    
+    # return service ipv6 address for dualstack cluster
+    def ip_v6(user: nil, cached: true, quiet: false)
+      spec = raw_resource(user: user, cached: cached, quiet: quiet).dig('spec')
+      ipv6 = spec.dig('clusterIPs')[0].include?(":") ? spec.dig('clusterIPs')[0] : spec.dig('clusterIPs')[1]
```

Comment:
Is there any documentation that describes IPv6 will always be the second IP? If this is not guaranteed, then both IPs need to be inspected and the correct one chosen. Same applies to Pod.

Reply:
I guess it should be according the order of network config:
```
  spec:
    clusterNetwork:
    - cidr: 10.128.0.0/14                --> ipv4 first
      hostPrefix: 23
    - cidr: fd01::/48                  --> then ipv6 
      hostPrefix: 64
    externalIP:
      policy: {}
    networkType: OVNKubernetes
    serviceNetwork:
    - 172.30.0.0/16
    - fd02::/112
  status:
    clusterNetwork:
    - cidr: 10.128.0.0/14
      hostPrefix: 23
    - cidr: fd01::/48
      hostPrefix: 64
    clusterNetworkMTU: 1400
    networkType: OVNKubernetes
    serviceNetwork:
    - 172.30.0.0/16             ##--> same here
    - fd02::/112

```

Reply:
I now see that you are checking which one is the correct. I would suggest to use need something like
```
ipv6 = spec['clusterIPs'].find { |ip| ip.include? "::" }
```

In this way you will get `nil` result in case IPv6 IP is missing.

It also makes sense to add an `ip_v4` method to return an IPv4 address with same premise - that it might be missing.

The `ip` method would return the default one - v4 or v6.

Reply:
thanks @akostadinov 
add ipv4 and ipv6 method to make sure get the correct ip address

Reply:
> I guess it should be according the order of network config:
> 
> ```
>   spec:
>     clusterNetwork:
>     - cidr: 10.128.0.0/14                --> ipv4 first
>       hostPrefix: 23
>     - cidr: fd01::/48                  --> then ipv6 
>       hostPrefix: 64
>     externalIP:
>       policy: {}
>     networkType: OVNKubernetes
>     serviceNetwork:
>     - 172.30.0.0/16
>     - fd02::/112
>   status:
>     clusterNetwork:
>     - cidr: 10.128.0.0/14
>       hostPrefix: 23
>     - cidr: fd01::/48
>       hostPrefix: 64
>     clusterNetworkMTU: 1400
>     networkType: OVNKubernetes
>     serviceNetwork:
>     - 172.30.0.0/16             ##--> same here
>     - fd02::/112
> ```

For Pod, first IP is always ipv4
Status:       Running
IP:           10.128.3.230
IPs:
  IP:           10.128.3.230
  IP:           fd01:0:0:5::1e6

For service, the order of IP is based on the the order of IP Families, ' IPv4,IPv6' or 'IPv6,IPv4', check below test logs:
$ oc create -f https://raw.githubusercontent.com/weliang1/Openshift_Networking/master/Features/Services/DualStack/PreferDualStack-v4v6-svc.yaml
$ oc describe svc preferdualstack-v4v6-svc
Type:              ClusterIP
IP Family Policy:  PreferDualStack
IP Families:       IPv4,IPv6
IP:                172.30.51.16
IPs:               172.30.51.16,fd02::a46f

$ oc create -f https://raw.githubusercontent.com/weliang1/Openshift_Networking/master/Features/Services/DualStack/PreferDualStack-v6v4-svc.yaml
$ oc describe svc preferdualstack-v6v4-svc
Type:              ClusterIP
IP Family Policy:  PreferDualStack
IP Families:       IPv6,IPv4
IP:                fd02::45c3
IPs:               fd02::45c3,172.30.206.223

$ oc get svc
NAME                       TYPE        CLUSTER-IP     EXTERNAL-IP   PORT(S)    AGE
preferdualstack-v4v6-svc   ClusterIP   172.30.51.16   <none>        8080/TCP   109s
preferdualstack-v6v4-svc   ClusterIP   fd02::45c3     <none>        8080/TCP   11s

Reply:
yes, I know service can set the order by `IP Families: IPv6,IPv4`
not sure what's the result for pod if we set ipv6 first, eg
```
    clusterNetwork:
    - cidr: fd01::/48                  -->  ipv6 
      hostPrefix: 64
    - cidr: 10.128.0.0/14                --> ipv4 
      hostPrefix: 23
```
but anyways, the current implementation can work for both

Reply:
IP Families is not for pod, pod always get both ip address in dual stack cluster.
IPs:
IP: 10.128.3.230
IP: fd01:0:0:5::1e6

Reply:
yes,  I know `IP Families: IPv6,IPv4` is only for service. 

I mean if we set ipv6 first in network config as below. 


```
clusterNetwork:
    - cidr: fd01::/48                  -->  ipv6 
      hostPrefix: 64
    - cidr: 10.128.0.0/14                --> ipv4 
      hostPrefix: 23
```
then pod ip may show: 
```
IPs:
IP: fd01:0:0:5::1e6
IP: 10.128.3.230
```